File: leetcode/15.三数之和.py
#
# @lc app=leetcode.cn id=15 lang=python3
#
# [15] 三数之和
#

# @lc code=start
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums = sorted(nums)
        ans = []
        for pos, x in enumerate(nums):
            if (pos+2 == len(nums)):
                break
            if (pos > 0 and nums[pos-1] == nums[pos]): continue
            if (x+nums[pos+1]+nums[pos+2] > 0): break
            if (x+nums[-1]+nums[-2] < 0): continue
            l, r = pos+1, len(nums)-1
            while (l < r):
                s = x + nums[l] + nums[r]
                if (s == 0):
                    ans.append([nums[pos], nums[l], nums[r]])
                    l += 1
                    while (l < r and nums[l] == nums[l-1]):
                        l += 1
                    r -= 1
                    while (l < r and nums[r] == nums[r+1]):
                        r -= 1
                elif (s > 0):
                    r -= 1
                else:
                    l += 1
        return ans
        
# @lc code=end

File: leetcode/15.三数之和.cpp
/*
 * @lc app=leetcode.cn id=15 lang=cpp
 *
 * [15] 三数之和
 */

// @lc code=start
#include <algorithm>
class Solution {
   public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        vector<vector<int>> ans;
        std::ranges::sort(nums);
        int len = nums.size();
        for (int a = 0; a < len-2; a++) {
            if (a > 0 && nums[a] == nums[a-1])
                continue;
            if (nums[a]+nums[a+1]+nums[a+2]>0)
                break;
            if (nums[a]+nums[len-2]+nums[len-1]<0)
                continue;
            int l = a + 1;
            int r = len - 1;
            while (l < r) {
                int x = nums[a] + nums[l] + nums[r];
                if (x == 0)
                {
                    ans.push_back({nums[a], nums[l++], nums[r--]});
                    while (l < r && nums[l] == nums[l-1]) l++;
                    while (l < r && nums[r] == nums[r+1]) r--;
                }
                else if (x > 0)
                    r--;
                else
                    l++;
            }
        }
        return ans;
    }
};
// @lc code=end
